### PR TITLE
test(progress): refactoring unit tests

### PR DESCRIPTION
**Pull request by kdinev**
Closes #  

### Additional information (check all that apply):
 - [ ] Bug fix
 - [ ] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [x] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/progressbar/progressbar.component.ts [hunk 1]
```diff
@@ -184,10 +184,6 @@ export abstract class BaseProgressDirective {
 
         this._internalState.newVal = Math.round(toValue(toPercent(this.value, maxNum), maxNum));
         this._value = this._internalState.oldVal = Math.round(toValue(this.valueInPercent, maxNum));
-        if (maxNum < this._value) {
```
Comment (hanastasov) [inline projects/igniteui-angular/src/lib/progressbar/progressbar.component.ts:187]: Setting the `max` runtime with a value lesser than current value would potentially enter this if. However, in such case, `this._value` will be recalculated wrongly to 0 on line 186. IMO we should place the `maxNum < this._value` check in the  the first if statement in the max setter.

**Review by hanastasov — CHANGES_REQUESTED** [review]

File: projects/igniteui-angular/src/lib/progressbar/progressbar.component.ts (hunk 1 above)
Comment (kdinev) [reply projects/igniteui-angular/src/lib/progressbar/progressbar.component.ts:187]: @hanastasov I have added some tests and they pass without this snippet. Have you tested that the logic doesn't work without it, or is it an assumption. Either something is wrong with the tests, or there's some other logic also normalizing the `_value` depending on the new `max`.

Comment (hanastasov) [reply projects/igniteui-angular/src/lib/progressbar/progressbar.component.ts:187]: I tested it and it seemed to lead the progressbar in wrong state. Looking at `Value should not exceed the max limit when operating with floating numbers` test now.

Comment (hanastasov) [reply projects/igniteui-angular/src/lib/progressbar/progressbar.component.ts:187]: OK, turns out what I mentioned is an edge case when max is set to 0. I have added a test that fails now and will add a fix in next commit.

Comment (hanastasov) [reply projects/igniteui-angular/src/lib/progressbar/progressbar.component.ts:187]: I've added a fix for the edge case I explained. When outside the edge case, the logic in `progressValue` normalizes the `this._value`, as you assumed.

**Review by hanastasov — DISMISSED** [review]

File: projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts [hunk 2]
```diff
@@ -45,405 +44,259 @@ describe('IgxLinearBar', () => {
...
+        fixture.detectChanges();
 
-            fix.detectChanges();
-            expect(bar.classList.contains(INDETERMINATE_CLASS)).toEqual(true);
-        });
+        expect(bar.classList.contains(INDETERMINATE_CLASS)).toEqual(true);
     });
 });
-
-@Component({ template: `<igx-linear-bar [animate]="false"></igx-linear-bar>` })
```
Comment (kdinev) [inline projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:431]: @Aleksandyr Take a look at the component before - animations were still off, so nothing has changed here.

Comment (Aleksandyr) [reply projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:431]: You are right.
I saw that 2 months ago I have disabled all animations when they were improved and that has been merged without noting that.
image
We need to change that because we miss several important cases when progress is updated with animation.
Especially when setting `max` and `value` runtime.

Comment (kdinev) [reply projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:431]: @Aleksandyr I actually tried to add a few tests with animations on, and since we use the native browser animations, it made the tests really unstable. I think this can be left for a more "UI test" scenario, while keeping the current test suite as unit tests only.

Comment (Aleksandyr) [reply projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:431]: I tried as well and it faced the same result as you. I am going to add an issue for UI animation tests

File: projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts [hunk 3]
```diff
@@ -45,405 +44,259 @@ describe('IgxLinearBar', () => {
...
-
         expect(progress.value).toBe(progressBarValue);
     });
 
-    it('Should update value when we try to decrese it', () => {
-        const fixture = TestBed.createComponent(LinearBarComponent);
-        fixture.detectChanges();
-        const progressBar = fixture.componentInstance.progressbar;
-        let expectedValue = 50;
+    it('when update step is bigger than passed value the progress indicator should follow the value itself', () => {
```
[the remark this answers is not in the document]
Comment (kdinev) [reply projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:138]: I looked through those tests and neither tests the `increment` functionality. All of them are testing logic which is applicable even without animations like setting a larger step than max value, which should revert to the previous step value, etc. I think these tests should stay.

File: projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts [hunk 4]
```diff
@@ -45,405 +44,259 @@ describe('IgxLinearBar', () => {
...
-        const fix = TestBed.createComponent(InitLinearProgressBarComponent);
-        fix.detectChanges();
-
-        const progressbar = fix.componentInstance.linearBar;
-        expect(progressbar.value).toEqual(0);
-
-        progressbar.indeterminate = true;
-        progressbar.value = 30;
-        fix.detectChanges();
+    it('Should change class suffix which would be relevant to the type that has been passed', () => {
```
[the remark this answers is not in the document]
Comment (kdinev) [reply projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:255]: There are getters in the code, which are host bound to the template. I can break each of them by changing the enumeration value the getters are comparing the `type` against. Seems like a unit test to me.

[the remark this answers is not in the document]
Comment (Aleksandyr) [reply projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:255]: Agree

File: projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Aleksandyr) [reply projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:138]: This test should stay as you suggested.

File: projects/igniteui-angular/src/lib/progressbar/circularbar.component.spec.ts [hunk 5]
```diff
@@ -11,449 +10,360 @@ const CIRCULAR_OUTER_CLASS = 'igx-circular-bar__outer';
...
+            // Should not set a step larger than max value
+            progress.step = 20;
+            expect(progress.step).toBe(step);
+            expect(progress.max).toBe(max);
+        });
 
-        expect(bar.value).toBe(expectedRes);
-        expect(bar.valueInPercent).toBe(expectedRes);
-    });
+        it(`when step value is not divisble to passed value the result returned from the value getter should be the same as the passed one`, () => {
```
Comment (Aleksandyr) [inline projects/igniteui-angular/src/lib/progressbar/circularbar.component.spec.ts:175]: I see that we test increment/decrement functionality which does not include the step property.

File: projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts [hunk 6]
```diff
@@ -45,405 +44,259 @@ describe('IgxLinearBar', () => {
...
 
-        expectedValue = 20;
-        fixture.componentInstance.value = expectedValue;
+    it('Prevent constant update of progress value when value and max value differ', () => {
+        const maxVal = 3.25;
+        const value = 2.55;
+        const progressBar = fixture.debugElement.nativeElement;
+
+        expect(progressBar.attributes['aria-valuenow'].textContent).toBe(progress.value.toString());
+        progress.step = 0.634;
```
Comment (Aleksandyr) [inline projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:175]: Passing this value doesn't affect the test. It will always be valid no matter the step.

Comment (kdinev) [reply projects/igniteui-angular/src/lib/progressbar/linearbar.component.spec.ts:175]: I have not changed this test from the previous batch.

**Comment by Aleksandyr** [comment]
I thought that we might apply some changes that might improve the test suite. But as we logged another issue for [that](https://github.com/IgniteUI/igniteui-angular/issues/10965) we might apply them there.

**Review by Aleksandyr — APPROVED** [review]

**Review by Aleksandyr — APPROVED** [review]
